hera_qm/xrfi.py

```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+# Functions for flagging auto-correlations
+# ############################################################################
+
+def dpss_flagger(data, freqs, filter_centers, filter_half_widths, wgts=None,
+                 nsig=10, mode="dpss_solve", noise=None, eigenval_cutoff=[1e-9],
+                 suppression_factors=[1e-9], cache={}):
+    """
+    Flag visibilities by filtering data with discrete prolate spheroidal sequences
+
+    Parameters:
+    ----------
+    data: np.ndarray
+        2D data array of the shape (time, frequency)
+    freqs: np.ndarray
+        1D array of frequencies present in the data
+    filter_centers: array-like
+        list of floats of centers of delay filter windows in nanosec
+    filter_half_widths: array-like
+        list of floats of half-widths of delay filter windows in nanosec
+    wgts: np.ndarray
+        2D np.ndarray of real weights. Must be the same shape as data.
+    nsig: float, default=10
+        The number of sigma in the metric above which to flag pixels.
+    mode: str, default='dpss_solve'
+        Method used to solve for DPSS model components. Options are 'dpss_matrix', 'dpss_solve', and 'dpss_leastsq'.
+    noise: np.ndarray, default=None
+        2D array for containing an estimate of the noise standard deviation. Must be the same shape as the data.
+        If noise=None, the noise standard deviation will be estimated from the filtered residuals.
+    eigenval_cutoff: array-like, default=[1e-9]
+        List of sinc_matrix eigenvalue cutoffs to use for included DPSS modes.
+    suppression_factors: array-like, default=[1e-9]
+        Specifies the fractional residuals of model to leave in the data. For example, 1e-6 means that the filter
+        will leave in 1e-6 of data fitted by the model.
+    cache: dictionary, default={}
+        dictionary for caching fitting matrices.
+
+    Returns:
+    -------
+    wgts: np.ndarray
+        Array containing weights for each antenna with values of either zero or one with
+        zero identifying channels found to have RFI
+    """
+    if len(suppression_factors) == 1 and len(filter_centers) > 1:
+        suppression_factors = len(filter_centers) * suppression_factors
+         
+    if len(eigenval_cutoff) == 1 and len(filter_centers) > 1:
+        eigenval_cutoff = len(filter_centers) * eigenval_cutoff
+
+    if wgts is None:
+        wgts = np.ones_like(data)
+
+    else:
+        wgts = deepcopy(wgts)
+
+    # Compute first-pass model and residuals
+    model, _, _ = dspec.fourier_filter(
+        freqs, data, wgts, filter_centers, filter_half_widths, mode=mode,
+        suppression_factors=suppression_factors, eigenval_cutoff=eigenval_cutoff,
+        cache=cache
+    )
+    res = data - model
+
+    # Compute modified z-score
+    sigma = np.median(
+        np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+    )
+    sigma /= 0.675
+    wgts = np.where(np.abs(res) < nsig * sigma, 1, 0)
```

Comment:
Why not use the input noise where you have it, instead of 1s?

Reply:
Changed to use input noise when possible